fetchcode/package_registry/__init__.py

```diff
@@ -0,0 +1,154 @@
+# fetchcode is a free software tool from nexB Inc. and others.
+# Visit https://github.com/nexB/fetchcode for support and download.
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# http://nexb.com and http://aboutcode.org
+#
+# This software is licensed under the Apache License version 2.0.
+#
+# You may not use this software except in compliance with the License.
+# You may obtain a copy of the License at:
+# http://apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+from attr import attrs, attrib
+
+from packageurl.contrib.route import NoRouteAvailable
+from packageurl import PackageURL
+from packageurl.contrib.route import Router
+import requests
+
+
+router = Router()
+
+
+@attrs
+class URLData:
+    """
+    Represent the Data extracted from any URL/PURL
+    `homepage_url` : URL to the homepage for this package
+    `api_url` :  Return the package repository API URL to obtain structured data for this package
+    `vcs_url` : URL for VCS download
+    `codeview_url` : URL where the code can be browsed online
+    `bugtracking_url` : URL to the issue or bug tracker for this package
+    `download_url` :  Return the package repository download URL to download the actual archive of code of this package. 
+    `documentation_url` : URL where the documentation can be found for this package
+    `readme_url` : URL where readme can be found for this package
+    `reverse_dependencies_url` : URL where reverse dependencies can be found for this package
+    `author_url` : URL where authors can be found regarding this package
+    `tags` : List of tags for this package
+    `tags_download_urls` : List of download URLS for tags of this package
+    `tags_readme_urls` : List of readme URLS for tags of this package
+    `tags_dependency_urls` : List of dependency URLs for tags of this package 
+    """
+    homepage_url = attrib(default=None)
+    api_url = attrib(default=None)
+    vcs_url = attrib(default=None)
+    codeview_url = attrib(default=None)
+    bugtracking_url = attrib(default=None)
+    download_url = attrib(default=None)
+    documentation_url = attrib(default=None)
+    readme_url = attrib(default=None)
+    reverse_dependencies_url = attrib(default=None)
+    author_url = attrib(default=None)
+    tags = attrib(default=None)
+    tags_download_urls = attrib(default=None)
+    tags_readme_urls = attrib(default=None)
+    tags_dependency_urls = attrib(default=None)
```

Comment:
In the fetchcode context, we may want to subclass this URLData object. For instance, an ftp url would not necessarily have all these fields (like `homepage_url`, `readme_url` etc). Some of these would be better off in a PackageURLData or similar and leaving the `URLData` object to be more of a generic object.

Reply:
Agreed 💯